model_hub/model_hub/huggingface/_trial.py

```diff
@@ -0,0 +1,256 @@
+    AutoModelForNextSentencePrediction,
+    AutoModelForTokenClassification,
+    AutoModelForQuestionAnswering,
+    AutoConfig,
+    AutoTokenizer,
+)
+from transformers.optimization import AdamW, Adafactor, get_scheduler
+from datasets import load_dataset
+
+from determined.pytorch import (
+    LRScheduler,
+    PyTorchTrial,
+    PyTorchTrialContext,
+)
+
+from ._arg_parser import (
+    default_parse_config_tokenizer_model_args,
+    default_parse_optimizer_lr_scheduler_args,
+    parse_dict_to_dataclasses,
+    OptimizerKwargs,
+    LRSchedulerKwargs,
+    DatasetKwargs,
+)
+
+
+MODEL_MODES = {
+    "base": AutoModel,
+    "pretraining": AutoModelForPreTraining,
+    "causal-lm": AutoModelForCausalLM,
+    "masked-lm": AutoModelForMaskedLM,
+    "seq2seq-lm": AutoModelForSeq2SeqLM,
+    "sequence-classification": AutoModelForSequenceClassification,
+    "multiple-choice": AutoModelForMultipleChoice,
+    "next-sentence": AutoModelForNextSentencePrediction,
+    "token-classification": AutoModelForTokenClassification,
+    "question-answering": AutoModelForQuestionAnswering,
+}
+
+
+def build_using_auto(
+    config_args: [Dict, AttrDict],
+    tokenizer_args: [Dict, AttrDict],
+    model_mode: str,
+    model_args: [Dict, AttrDict],
+):
+    """
+    Build the config, tokenizer, and model using tranformer's
+    Auto classes.
+
+    Args:
+        config_args: arguments for transformers configuration classes
+        tokenizer_args: arguments for transformers tokenizer classes
+        model_mode: one of the tasks supported by transformers, see MODEL_MODES for
+            the supported options
+        model_args: arguments for transformers model classes
+
+    Returns:
+        transformer config, tokenizer, and model
+    """
+    config = AutoConfig.from_pretrained(**config_args)
+    tokenizer = AutoTokenizer.from_pretrained(**tokenizer_args)
+    model_mode = MODEL_MODES[model_mode]
+    model_args["config"] = config
+    model = model_mode.from_pretrained(**model_args)
+    return config, tokenizer, model
+
+
+def group_parameters_for_optimizer(
+    model: torch.nn.Module, weight_decay: float, no_decay=("bias", "LayerNorm.weight")
+):
+    """
+    Group parameters by whether weight_decay is applied or not.
+
+    Args:
+        model: model supplying the learnable parameters
+        weight_decay: value for weight_decay
+        no_decay: variable names that should not have weight_decay applied
+    Returns:
+        grouped parameters according to whether weight_decay should be applied
+    """
+    return [
+        {
+            "params": [
+                p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)
+            ],
+            "weight_decay": weight_decay,
+        },
+        {
+            "params": [p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)],
+            "weight_decay": 0.0,
+        },
+    ]
+
+
+def build_default_optimizer(
```

Comment:
Does Huggingface only use Adafactor or AdamW?

Reply:
yes, this matches the behavior of the HF trainer if no optimizer is provided.
https://github.com/huggingface/transformers/blob/master/src/transformers/trainer.py#L616